### DISCUSSION TITLE
How to use the method signalMsgConsumer on Peppol client

### DISCUSSION THREAD

**Discussion by luukth-infosupport** [phax/phase4#119] [Q&A] [answered]
Hi, I have a question regarding the signalMsgConsumer. Why is using this callback to only way to get detailed information about the response of the sent document? I would assume that the signalMessage would be a response of calling the sendMessageAndCheckForReceipt method, but instead it only returns an Enum which doesn't give a lot of detailed information.
```java
ESimpleUserMessageSendResult receipt =
                Phase4PeppolSender.builder()
                        .documentTypeID(documentTypeIdentifier)
                        .processID(processIdentifier)
                        .senderParticipantID(senderParticipantIdentifier)
                        .receiverParticipantID(receiverParticipantIdentifier)
                        .fromPartyID(fromPartyId)
                        .smpClient(smpClient)
                        .payload(...)
                        .signalMsgConsumer(...)
                        .sendMessageAndCheckForReceipt();
```
What would be the recommended approach to get signalMessage information into the current "context". Right now I'm using a `CompletableFuture<Ebms3SignalMessage>` and calling this `signalMsgConsumer(future::complete)`. After that I use the following code to get the value. Are there alternative ways to receive a similar result?
```java
if (future.isDone()) {
    response.ebms3SignalMessage = future.join();
}
```

**Comment by phax** [comment, marked as answer]
Hi @luukth-infosupport,
The reasons for this rather "difficult" flow to retrieve Signal Messages are quite simple
* A lot can go wrong - prior to sending, during sending (on a network and HTTP level) and after sending (response handling) - the subtleties of these changes do matter to callers. Just returning the SignalMessage or `null` would not help to spot any error
* The SignalMessage itself contains no "content" - it is usually just an acknowledgement on the transport level. The only main "content" is the non-repudiation of receipt which is effectievely a bunch of XMLDsig like structures. Therefore the SignalMessage is not of so much relevance on a processing level (of course you need to store it etc.)

My proposed approach to just store the SignalMessage is like this:
```java
com.helger.commons.wrapper.Wrapper<Ebms3SignalMessage> signalMessageHolder = new com.helger.commons.wrapper.Wrapper<> ();
SimpleUserMessageSendResult receipt =
                Phase4PeppolSender.builder()
                       // ...
                        .signalMsgConsumer(signalMessageHolder::set)
                        .sendMessageAndCheckForReceipt();
// the above call is synchronous
Ebms3SignalMessage receivedSignalMsg = signalMessageHolder.get ();
```

hth
